File: backend/app/models/document.py
from datetime import datetime
from typing import List, Optional, Dict, Annotated
from pydantic import BaseModel, Field, BeforeValidator, PlainSerializer
from bson import ObjectId


def validate_object_id(v):
    """Validate and convert to ObjectId"""
    if isinstance(v, ObjectId):
        return v
    if ObjectId.is_valid(v):
        return ObjectId(v)
    raise ValueError("Invalid ObjectId")


# Pydantic v2 compatible ObjectId type
PyObjectId = Annotated[
    ObjectId,
    BeforeValidator(validate_object_id),
    PlainSerializer(lambda x: str(x), return_type=str)
]


class Statistic(BaseModel):
    name: str
    value: float
    unit: Optional[str] = None
    visualization_type: str = "bar"  # bar, line, pie, gauge


class DesignSpecification(BaseModel):
    background_color: str = "#FFFFFF"
    foreground_color_1: str = "#2563EB"
    foreground_color_2: str = "#06B6D4"
    theme_name: Optional[str] = "Ocean Blue"


class DocumentConfig(BaseModel):
    length: int  # Target word count
    document_type: str = "infographic"  # formal, infographic, or invoice
    use_watermark: bool = False  # Only applicable for formal documents with logo
    statistics: List[Statistic] = []
    design_spec: DesignSpecification


class DocumentFiles(BaseModel):
    logo_url: Optional[str] = None
    pdf_url: Optional[str] = None
    generated_images: List[str] = []
    visualizations: List[str] = []


class GenerationMetadata(BaseModel):
    text_model: str
    image_model: str
    word_count: int = 0
    generation_time_seconds: float = 0.0


class Document(BaseModel):
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    user_id: str = "default"
    title: str
    description: str
    status: str = "pending"  # pending, processing, completed, failed
    config: DocumentConfig
    files: DocumentFiles = Field(default_factory=DocumentFiles)
    generation_metadata: Optional[GenerationMetadata] = None
    error_message: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: Optional[datetime] = None

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}


class GenerationJob(BaseModel):
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    document_id: PyObjectId
    user_id: str = "default"  # User who owns this job
    status: str = "pending"  # pending, processing, completed, failed
    progress: int = 0  # 0-100
    current_step: str = "initializing"
    error_message: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}

File: frontend/src/ts/types/document.ts
export interface Statistic {
  id: string;
  name: string;
  value: number;
  unit?: string;
  visualization_type: 'bar' | 'line' | 'pie' | 'gauge';
}

export interface DesignSpecification {
  background_color: string;
  foreground_color_1: string;
  foreground_color_2: string;
  theme_name?: string;
}

export interface DocumentGenerationRequest {
  description: string;
  length: number;
  document_type: 'formal' | 'infographic' | 'invoice';
  use_watermark: boolean;  // Only applicable for formal documents with logo
  statistics: Statistic[];
  design_spec: DesignSpecification;
  logo?: File;
}

export interface GenerationJobResponse {
  job_id: string;
  status: string;
  message: string;
  estimated_time_seconds: number;
}

export interface JobStatusResponse {
  job_id: string;
  document_id?: string;
  status: string;
  progress: number;
  current_step: string;
  error_message?: string;
}

export interface DocumentResponse {
  id: string;
  title: string;
  description: string;
  status: string;
  created_at: string;
  updated_at: string;
  completed_at?: string;
  pdf_url?: string;
}

export interface ColorTheme {
  name: string;
  background: string;
  foreground1: string;
  foreground2: string;
}

export const COLOR_THEMES: ColorTheme[] = [
  {
    name: 'Ocean Blue',
    background: '#FFFFFF',
    foreground1: '#2563EB',
    foreground2: '#06B6D4',
  },
  {
    name: 'Corporate Red',
    background: '#FFFFFF',
    foreground1: '#DC2626',
    foreground2: '#F97316',
  },
  {
    name: 'Forest Green',
    background: '#FFFFFF',
    foreground1: '#059669',
    foreground2: '#14B8A6',
  },
  {
    name: 'Royal Purple',
    background: '#FFFFFF',
    foreground1: '#7C3AED',
    foreground2: '#EC4899',
  },
  {
    name: 'Sunset Orange',
    background: '#FFFFFF',
    foreground1: '#EA580C',
    foreground2: '#EAB308',
  },
];
